clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerClientInitialResponse.java

```diff
@@ -97,21 +132,41 @@ public SaslExtensions extensions() {
         return message.getBytes(StandardCharsets.UTF_8);
     }
 
+    /**
+     * Return the always non-null token value
+     * 
+     * @return the always non-null toklen value
+     */
     public String tokenValue() {
         return tokenValue;
     }
 
+    /**
+     * Return the always non-null authorization ID
+     * 
+     * @return the always non-null authorization ID
+     */
     public String authorizationId() {
         return authorizationId;
     }
 
     /**
      * Validates that the given extensions conform to the standard. They should also not contain the reserve key name {@link OAuthBearerClientInitialResponse#AUTH_KEY}
      *
+     * @param extensions
+     *            optional extensions to validate
+     * @return always non-null (potentially empty) extensions
+     * @throws SaslException
+     *             if any extension name or value fails to conform to the required
+     *             regular expression as defined by the specification, or if the
+     *             reserved {@code auth} appears as a key
+     *
      * @see <a href="https://tools.ietf.org/html/rfc7628#section-3.1">RFC 7628,
      *  Section 3.1</a>
      */
     public static SaslExtensions validateExtensions(SaslExtensions extensions) throws SaslException {
+        if (extensions == null)
```

Comment:
What do you think about ensuring that non-null SaslExtensions never enter the system, rather than sprinkling checks in the code?

Since `SaslExtensions` are passed by `SaslExtensionsCallback`, I think it would be simple to guarantee no `null` extensions are ever returned by its `extensions()` method.

Reply:
This is the only place we check if it is null as far as I can tell.  If we had to check in multiple places then it would be worth addressing.  I added the Javadoc to make it clear that null is acceptable as an input but null is never returned.  I'm not 100% comfortable with that -- requiring non-null as input and returning the input would be better.  But now that I write that, it occurs to me -- why does this method return anything at all?  A better signature might be "void confirmValid(SaslExtensions) throws SaslException" and it would throw an exception if it weren't valid -- and accept null as being valid.  The only place the return value is used is in the constructor for this class, and we can easily check for null there.

Reply:
That's how it was initially - here's your suggestion to have it return the extensions - https://github.com/apache/kafka/pull/5379/commits/dd9b15b61254d58bb26423b1548aaba5b0ba6bb5#r204512923

Even if this is the only place, I still feel like it might be better to fix it at the source rather than the destination where it is used. I don't know if we'll have any other place where we'll use them even in the future (third client callback, perhaps?) but it just sounds cleaner to me that way. Maybe we can have somebody else share their opinion

Reply:
<<<That's how it was initially - here's your suggestion to have it return the extensions
Yes, I can see how that might be confusing.  To clarify, at https://github.com/apache/kafka/pull/5379/files/4e4ac78f03a200f46bfe31cf466cdf0dc17d2b30#diff-3de90af12234c40ce1647059e6725a55R81 I did suggest an improvement at that time: the constructor was invoking the method `void extensions(SaslExtensions)`, which was setting `this.extensions` after invoking `void validateExtension(SaslExtensions)`.  The main motivation I had for that suggestion was to set `this.extensions` in the constructor instead of elsewhere.  I did also suggest that `void validateExtensions(SaslExtensions)` instead return a value so we could write `this.extensions = validateExtensions(extensions)` in the constructor, and I can see now that part of the suggestion -- making `validateExtensions()` return a value -- was likely a mistake.  I think the suggestion I am making now would fix that -- it keeps the assignment of `this.extensions` in the constructor where it belongs while keeping the semantics of the method `validateExtensions()` simple (or `confirmValid()`, or whatever we call it if we change it -- I don't have a strong feeling about the name).  The semantics of the method, as per the Javadoc, is to make sure the names and values conform to the specification's regular expression requirements and make sure the reserved "auth" name doesn't appear as an extension key.  Currently as implemented `SaslExtensions validateExtensions(SaslExtensions)` complects 2 things: it validates what is passed in for legality, but it also takes on the responsibility of returning something that is valid: that second responsibility means having to deal with null to either return null or return an empty `SaslExtensions` instance.  The method should be simple and do 1 thing only, which is to validate that what is given is legal.  Then whoever calls the method, once they know what they have is legal, can determine what they want to do with it (i.e. maybe they want to keep null, or maybe they don't).  I don't think we want to say "null is illegal" in this case; I think it is legal for backwards compatibility.  So I think this method should be `void validateExtension(SaslExtensions)` with null being considered legal, and then it is up to the constructor to store NO_SASL_EXTENSIONS in its `this.extensions` field in place of null.

Does that clarify it?

Reply:
>  I don't think we want to say "null is illegal" in this case; I think it is legal for backwards compatibility. 

I don't see any backwards compatibility in place here - that method would be introduced in `2.1.0` and this PR too.

I do agree that it is better for `validateExtensions()` to do one thing only and that is validate the extensions without returning anything.

The question of this comment thread is: where should the potential `null` value be handled?
My thoughts, in order of preference, are:
1. `SaslExtensionsCallback` - ensure it doesn't return null
2. `OAuthBearerInitialClientResponse`'s constructor
3. `validateExtensions()` method

Reply:
`SaslExtensionsCallback` might be unhandled, and in that case it is going to return null.  I think that is where backwards compatibility becomes relevant.  I'm going to push a commit that implements (2) above (handle null in `OAuthBearerInitialClientResponse`'s constructor); the commit will also eliminate the return value from `validateExtensions()`.  We can see what it looks like and back it out if necessary, but I think that's the way to go.  Thanks for the discussion -- it was fruitful, and we can continue as necessary.